crates/flox-rust-sdk/src/models/environment/managed_environment.rs

```diff
@@ -97,3 +146,700 @@ impl Environment for ManagedEnvironment {
+    ///
+    /// The definition of a managed environment is stored on a branch in a
+    /// central clone of the environment owner's floxmeta repository located in
+    /// `$FLOX_DATA_DIR/meta/<owner>`. Every .flox directory will correspond to
+    /// a unique branch.
+    ///
+    /// To open an environment at a given location:
+    ///
+    /// - We open a user's floxmeta clone, i.e. `$FLOX_DATA_DIR/meta/<owner>`.
+    ///   If that repo doesn't exist, it will be cloned.
+    ///
+    /// - We open the lockfile and ensure that the specific commit referenced is
+    ///   present in floxmeta. If it is not, we fetch the environment and lock
+    ///   to `HEAD`.
+    ///
+    /// - We check whether a unique branch for the .flox directory exists in
+    ///   floxmeta and points at the commit in the lockfile, creating or
+    ///   resetting the correct branch if necessary.
+    ///
+    /// At some point, it may be useful to create a ManagedEnvironment without
+    /// fetching or cloning. This would be more correct for commands like delete
+    /// that don't need to fetch the environment.
+    pub fn open(
+        flox: &Flox,
+        pointer: ManagedPointer,
+        dot_flox_path: impl AsRef<Path>,
+    ) -> Result<Self, EnvironmentError2> {
+        let floxmeta =
+            FloxmetaV2::open(flox, &pointer).map_err(ManagedEnvironmentError::OpenFloxmeta)?;
+
+        let lock = Self::ensure_locked(flox, &pointer, &dot_flox_path, &floxmeta)?;
+        Self::ensure_branch(
+            &branch_name(&flox.system, &pointer, &dot_flox_path),
+            &lock,
+            &floxmeta,
+        )?;
+
+        Ok(ManagedEnvironment {
+            _path: dot_flox_path.as_ref().to_path_buf(),
+            _system: flox.system.clone(),
+            _floxmeta: floxmeta,
+            _pointer: pointer,
+        })
+    }
+
+    /// Ensure:
+    /// - a lockfile exists
+    /// - the commit in the lockfile (`local_rev` or `rev``) exists in floxmeta
+    ///
+    /// This may perform a fetch.
+    fn ensure_locked(
+        flox: &Flox,
+        pointer: &ManagedPointer,
+        dot_flox_path: impl AsRef<Path>,
+        floxmeta: &FloxmetaV2,
+    ) -> Result<GenerationLock, EnvironmentError2> {
+        let lock_path = dot_flox_path.as_ref().join(GENERATION_LOCK_FILENAME);
+        let maybe_lock: Option<GenerationLock> = match fs::read(&lock_path) {
+            Ok(lock_contents) => Some(
+                serde_json::from_slice(&lock_contents)
+                    .map_err(ManagedEnvironmentError::InvalidLock)?,
+            ),
+            Err(err) => match err.kind() {
+                io::ErrorKind::NotFound => None,
+                _ => Err(EnvironmentError2::ReadManifest(err))?,
+            },
+        };
+
+        Ok(match maybe_lock {
+            // local_rev is Some
+            Some(
+                lock @ GenerationLock {
+                    rev: _,
+                    local_rev: Some(_),
```

Comment:
I couldn't figure out a way to bind to local_rev without also borrowing lock, but is there a way to do that?

Reply:
Well I was wondering if I could destructure with `Some(local_rev)` instead of `Some(_)` and then use that in the match body (`local_rev` instead of `lock.local_rev.as_ref().unwrap()`). But if not then yeah match guards are better

Reply:
Actually it's not that you're borrowing `lock` it's that when you assign it a name like that you're _moving_ the `GenerationLock` into `lock`, and at the same time you're also trying to move the `local_rev` field into the `foo` in `Some(foo)`. This is tricky because you're trying to also return `lock` so you can't return a partially moved value.

Instead what you can do is take a shared reference to the `GenerationLock` when destructuring via `ref lock @ ...` then you can also destructure the `Some(local_rev)` field. You still need to return `lock` though, so you'd have to clone it since you can't move it out of the shared reference. All in all the match guards seems easiest.

Reply:
Yeah match guards are best